https://issues.redhat.com/browse/ACM-23917 MCOA (#8364)
open-cluster-management/rhacm-docs: observability/obs_mcoa_config_apis.adoc

[#config-apis-mcoa]
= Configuring APIs for the multicluster obsservability add-on

> Configure the default metrics specific APIs of the multicluster observability add-on. When you enable the multicluster observability add-on, placements are automatically created for your managed clusters.

Reviewer: We don't create placements. We only use the global placement that already exist by default. But we autimatically add the default configuration resources to this placement in the ClusterManagementAddOn.

> When you reference a new placement in your `ClusterManagementAddon` resource, the `multicluster-observability-addon-manager` creates a specific default `PrometheusAgent` resource and adds it to the referenced placement. 

Reviewer: We also add the required ScrapeConfigs and PrometheusRules.

*Required access:* Cluster administrator

.Prerequisites

- You have installed and enabled the multicluster observability add-on. For more information, see xref:../observability/obs_enable_mcoa.adoc#enable-mcoa[Enabling the multicluster observability add-on].

Complete the following to configure the APIs for the multicluster observability add-on:

> . Configure the `PrometheusAgent` resource by defining the prometheus server for federating metrics from the default scrape interval. Complete the following steps:

Reviewer: Let's discuss this part tomorrow. There are things to change.

[#config-apis-mcoa]
= Configuring APIs for the multicluster observability add-on

Configure the default metrics-specific APIs of the multicluster observability add-on. When you reference a new placement in your `ClusterManagementAddon` resource, the `multicluster-observability-addon-manager` automatically creates specific default `PrometheusAgent` resources. The `multicluster-observability-addon-manager` adds the `PrometheusAgent` resources reference in the related `Placement` configurations.

> While there is one `PrometheusAgent` resource created by the placement, the default `ScrapeConfigs` and `PrometheusRules` are common to all placements. The following resources are the default configuration resources for the multicluster observability add-on: `PrometheusAgent, ScrapeConfigs`, and `PrometheusRules`.

Reviewer 2: ```suggestion
While there is one `PrometheusAgent` resource created by the placement, the default `ScrapeConfigs` and `PrometheusRules` are common to all placements. The following resources are the default configuration resources for the multicluster observability add-on: `PrometheusAgent`, `ScrapeConfigs`, and `PrometheusRules`.
```

*Required access:* Cluster administrator

.Prerequisites

- You have installed and enabled the multicluster observability add-on. For more information, see xref:../observability/obs_enable_mcoa.adoc#enable-mcoa[Enabling the multicluster observability add-on].

> 

Reviewer 2: ```suggestion

.Procedure

```

. Configure the `ScrapeConfig` resource to define a set of metrics for independent federation from Prometheus of your managed clusters. Complete the following steps:

.. Add the name of the job that you want to reference for the `jobName` parameter.
.. To ensure that Prometheus federates metrics, add the `/federate` URL path for the `metricsPath` parameter.
.. Add the metric name and labels that you want to collect. See the following YAML file example where the `ScrapeConfig` resource collects the `up` metric:

+
[source,yaml]
----
apiVersion: monitoring.rhobs/v1alpha1
kind: ScrapeConfig
metadata:
  name: some-metrics-to-collect
  namespace: open-cluster-management-observability
  labels:
    - app.kubernetes.io/component: <platform-metrics-collector> or <user-workload-metrics-collector>
spec:
  jobName: some-job-name
  metricsPath: /federate
  params:
    match[]:
    - '{__name__="up"}'
----

. Configure the `PrometheusRule` resource to limit the cardinality of your collected metrics on your hub cluster. Complete the following steps:

> .. Define alerting and recording rules for custom workloads on your managed clusters.

Reviewer: They can be defined for both platform and uwl monitoring. 

As stated in my document:

------

For Prometheus rules targeting user-workloads, you must include the specific annotation observability.open-cluster-management.io/target-namespace to designate the namespace where the rule should be deployed.

Failure to include this annotation will result in the rules being deployed to the default installation namespace. This default namespace contains the openshift.io/cluster-monitoring=true label, which means the rules will not be monitored by the in-cluster user-workload monitoring stack.

-----

Author: Added within the notes. Let me know what you think

> .. To define the namespace where you want to deploy your `PrometheusRule` resource, add the following annotation: `observability.open-cluster-management.io/target-namespace`.

Reviewer: We should explain why I think, as described in my above comment.

+
*Notes:* 

- By default, rules are not monitored by the user workload monitoring stack of your {ocp-short} cluster.
- Be sure to use the `monitoring.coreos.com` group for the `PrometheusRule` resource.

. Configure the `AddonDeploymentConfig` resource to customize the deployment of the multicluster observability add-on on managed clusters. *Note:* The values in the `AddonDeploymentConfig` resource are override direct modifications of other resources. Complete the following steps:

> .. Define the namespace where you install the `PrometheusAgent` resource for the `agentInstallNamespace` parameter.

Author: ```suggestion
.. Define the namespace where you install the `PrometheusAgent` resource for the `agentInstallNamespace` parameter. The default namespace is `open-cluster-management-agent-addon`.
```

.. To ensure that Prometheus federates metrics, add the `/federate` URL path for the `metricsPath` parameter.
.. Add the metric name and labels that you want to collect. See the following YAML file example where the `ScrapeConfig` resource collects the `up` metric:

+
[source,yaml]
----
apiVersion: monitoring.rhobs/v1alpha1
kind: ScrapeConfig
metadata:
  name: some-metrics-to-collect
  namespace: open-cluster-management-observability
  labels:
    - app.kubernetes.io/component: <platform-metrics-collector> or <user-workload-metrics-collector>
spec:
  jobName: some-job-name
  metricsPath: /federate
  params:
    match[]:
    - '{__name__="up"}'
----

. Configure the `PrometheusRule` resource to limit the cardinality of your collected metrics on your hub cluster. Complete the following steps:

.. Define alerting and recording rules for platform and user workload monitoring on your managed clusters. 

.. To target user workloads in your `PrometheusRule` resource, add the following annotation to define the namespace where you want to deploy the resource: `observability.open-cluster-management.io/target-namespace`. 

+
*Notes:* 

> - When the `observability.open-cluster-management.io/target-namespace` is not set in your `PrometheusRule` resource, the `PrometheusRule` resource are deployed to the default installation namespace. If `openshift.io/cluster-monitoring` is set to `true`, the `PrometheusRule` resouces are monitored by {ocp-short} user workload monitoring stack.

Reviewer: ```suggestion
- When the `observability.open-cluster-management.io/target-namespace` is not set in your `PrometheusRule` resource, the `PrometheusRule` resource are deployed to the default installation namespace. If `openshift.io/cluster-monitoring` is set to `true`, the `PrometheusRule` resouces are not monitored by {ocp-short} user workload monitoring stack.
```

[#config-apis-mcoa]
= Configuring APIs for the multicluster observability add-on

> Configure the default metrics-specific APIs of the multicluster observability add-on. When you reference a new placement in your `ClusterManagementAddon` resource, the `multicluster-observability-addon-manager` automatically creates specific default configuration resources and creates the configuration resources to the referenced placement. The following resources are the default configuration resources for the multicluster observability add-on: `PrometheusAgent, ScrapeConfigs`, and `PrometheusRules`.

Reviewer: ```suggestion
Configure the default metrics-specific APIs of the multicluster observability add-on. When you reference a new placement in your `ClusterManagementAddon` resource, the `multicluster-observability-addon-manager` automatically creates specific default `PrometheusAgent` resources. The `multicluster-observability-addon-manager` adds the `PrometheusAgent` resources reference in the related `Placement` configurations.

While there is one `PrometheusAgent` resource created by the placement, the default `ScrapeConfigs` and `PrometheusRules` are common to all placements. The following resources are the default configuration resources for the multicluster observability add-on: `PrometheusAgent, ScrapeConfigs`, and `PrometheusRules`.
```

*Required access:* Cluster administrator

.Prerequisites

- You have installed and enabled the multicluster observability add-on. For more information, see xref:../observability/obs_enable_mcoa.adoc#enable-mcoa[Enabling the multicluster observability add-on].

Complete the following to configure the APIs for the multicluster observability add-on:

. *Optional* Override the default scrape interval for your `PrometheusAgent` resource by changing the `scrapeInterval` parameter. The default value is `300s`. You can also override the `scrapeInterval` of the `scrapeConfig` resource.

> . Configure the `ScrapeConfig` resource to define a set of metrics for independent federation from Prometheus of the hub cluster. Complete the following steps:

Author: @thibaultmg mentioning here bc this is where we stopped before the meeting ended

> . Configure the `ScrapeConfig` resource to define a set of metrics for independent federation from Prometheus of the hub cluster. Complete the following steps:

Reviewer: ```suggestion
. Configure the `ScrapeConfig` resource to define a set of metrics for independent federation from Prometheus of your managed clusters. Complete the following steps:
```

.. Add the name of the job that you want to reference for the `jobName` parameter.
.. To ensure that Prometheus federates metrics, add the `/federate` URL path for the `metricsPath` parameter.
.. Add the metric name and labels that you want to collect. See the following YAML file example where the `ScrapeConfig` resource collects the `up` metric:

+
[source,yaml]
----
apiVersion: monitoring.rhobs/v1alpha1
kind: ScrapeConfig
metadata:
  name: some-metrics-to-collect
  namespace: open-cluster-management-observability
  labels:
    - app.kubernetes.io/component: <platform-metrics-collector> or <user-workload-metrics-collector>
spec:
  jobName: some-job-name
  metricsPath: /federate
  params:
    match[]:
    - '{__name__="up"}'
----

> .. Define the target prometheus source server for platform monitring and user workloads. Your `ScrapeConfig` resource might resemble the following YAML file example:

Reviewer: The user can only customized this target for UWL monitoring. We set a default pointing to the OCP uwl prometheus.
For platform monitoring, the user must not update these fields, they are enforced by MCOA.

+

> [source,yaml]

Reviewer: I would remove this.

----
apiVersion: monitoring.rhobs/v1alpha1
kind: ScrapeConfig
spec:
  ...
  scrapeClass: ocp-monitoring
  scheme: HTTPS
  staticConfigs:
  - targets:
    - prometheus-k8s.openshift-monitoring.svc:9091
    - prometheus-user-workload.openshift-user-workload-monitoring.svc:9092
----

. Configure the `PrometheusRule` resource to define alerting and recording rules for custom workloads on managed clusters.

+

> *Note:* Be sure to use the `monitoring.coreos.com` group for the `PrometheusRule` resource.

Reviewer: I think we should have a specific document for configuring rules. Or we should at least get this content here: https://docs.google.com/document/d/1joCg1kN4yd6HZ_GzJTRYXjNPvJGz_p4jhw5r26zYCI4/edit?tab=t.0#heading=h.5abtripm1var
There are specifics for uwl to be described. And limitations.

Author: If there are limitations, it should be added to the known issues for Observability. A new issue must be created to add new known issues.

Author: What parts of your draft need to be mentioned?

Author: Made a couple updates

. Configure the `AddonDeploymentConfig` resource to customize the deployment of the multicluster observability add-on on managed clusters. *Note:* The values in the `AddonDeploymentConfig` resource are override direct modifications of other resources. Complete the following steps:

> .. Define the namespace where you install the `PrometheusAgent` resource for the `agentInstallNamespace` parameter.

Reviewer: We can add that by default it is `open-cluster-management-agent-addon`

.. Add the name of the job that you want to reference for the `jobName` parameter.
.. To ensure that Prometheus federates metrics, add the `/federate` URL path for the `metricsPath` parameter.
.. Add the metric name and labels that you want to collect. See the following YAML file example where the `ScrapeConfig` resource collects the `up` metric:

+
[source,yaml]
----
apiVersion: monitoring.rhobs/v1alpha1
kind: ScrapeConfig
metadata:
  name: some-metrics-to-collect
  namespace: open-cluster-management-observability
  labels:
    - app.kubernetes.io/component: <platform-metrics-collector> or <user-workload-metrics-collector>
spec:
  jobName: some-job-name
  metricsPath: /federate
  params:
    match[]:
    - '{__name__="up"}'
----

.. Define the target prometheus source server for platform monitring and user workloads. Your `ScrapeConfig` resource might resemble the following YAML file example:

+
[source,yaml]
----
apiVersion: monitoring.rhobs/v1alpha1
kind: ScrapeConfig
spec:
  ...
  scrapeClass: ocp-monitoring
  scheme: HTTPS
  staticConfigs:
  - targets:
    - prometheus-k8s.openshift-monitoring.svc:9091
    - prometheus-user-workload.openshift-user-workload-monitoring.svc:9092

> ----

Author: ```suggestion
```
